scanpy/metrics/_morans_i.py

```diff
@@ -0,0 +1,263 @@
+    use_graph: Optional[str] = None,
+    layer: Optional[str] = None,
+    obsm: Optional[str] = None,
+    obsp: Optional[str] = None,
+    use_raw: bool = False,
+) -> Union[np.ndarray, float]:
+    r"""
+    Calculate Moran’s I Global Autocorrelation Statistic.
+
+    Moran’s I is a global autocorrelation statistic for some measure on a graph. It is commonly used in
+    spatial data analysis to assess autocorrelation on a 2D grid. It is closely related to Geary's C,
+    but not identical. More info can be found `here`<https://en.wikipedia.org/wiki/Moran%27s_I> .
+
+    .. math::
+
+        I=\frac{n}{S_{0}} \frac{\sum_{i=1}^{n} \sum_{j=1}^{n} w_{i, j} z_{i} z_{j}}{\sum_{i=1}^{n} z_{i}^{2}}
+
+    Params
+    ------
+    adata
+    vals
+        Values to calculate Moran's I for. If this is two dimensional, should
+        be of shape `(n_features, n_cells)`. Otherwise should be of shape
+        `(n_cells,)`. This matrix can be selected from elements of the anndata
+        object by using key word arguments: `layer`, `obsm`, `obsp`, or
+        `use_raw`.
+    use_graph
+        Key to use for graph in anndata object. If not provided, default
+        neighbors connectivities will be used instead.
+    layer
+        Key for `adata.layers` to choose `vals`.
+    obsm
+        Key for `adata.obsm` to choose `vals`.
+    obsp
+        Key for `adata.obsp` to choose `vals`.
+    use_raw
+        Whether to use `adata.raw.X` for `vals`.
+
+
+    This function can also be called on the graph and values directly. In this case
+    the signature looks like:
+
+    Params
+    ------
+    g
+        The graph
+    vals
+        The values
+
+
+    See the examples for more info.
+
+    Returns
+    -------
+    If vals is two dimensional, returns a 1 dimensional ndarray array. Returns
+    a scalar if `vals` is 1d.
+
+
+    Examples
+    --------
+
+    Calculate Morans I for each components of a dimensionality reduction:
+
+    .. code:: python
+
+        import scanpy as sc, numpy as np
+
+        pbmc = sc.datasets.pbmc68k_processed()
+        pc_c = sc.metrics.morans_i(pbmc, obsm="X_pca")
+
+
+    It's equivalent to call the function directly on the underlying arrays:
+
+    .. code:: python
+
+        alt = sc.metrics.morans_i(pbmc.obsp["connectivities"], pbmc.obsm["X_pca"].T)
+        np.testing.assert_array_equal(pc_c, alt)
+    """
+    if use_graph is None:
+        # Fix for anndata<0.7
+        if hasattr(adata, "obsp") and "connectivities" in adata.obsp:
+            g = adata.obsp["connectivities"]
+        elif "neighbors" in adata.uns:
+            g = adata.uns["neighbors"]["connectivities"]
+        else:
+            raise ValueError("Must run neighbors first.")
+    else:
+        raise NotImplementedError()
+    if vals is None:
+        vals = _get_obs_rep(adata, use_raw=use_raw, layer=layer, obsm=obsm, obsp=obsp).T
+    return morans_i(g, vals)
+
+
+@njit(
+    ft(ft[:], it[:], it[:], ft[:], ft[:], ft, ft),
```

Comment:
Why use static typing here? Why not just let numba generate specializations?

Reply:
mmh only preference is for:
- clarity
- make sure there is nothing weird happening down the line, as types are alwasy consistent (and conversion always happes).

If you have strong dislike I can of course remove

Reply:
So, most of the time the indices are going to be 32 bit (scipy tries to use a small width integer for these), and I do think it's important for the results here to be calculated with 64 bit accuracy since there can be a lot of operations per result value. However, some windows users have 32 bit systems, so locking to 64 bit can be problematic (see numba related issues around `calculate_qc_metrics`).

I'd say this should cast floating point data to system width, and be agnostic about the types of indices. That's what I did for `gearys_c`, where it looks like (from the comments) I came across precision problems with 32 bit values.

Reply:
ok but I need to stick to np.float32 for values, becuase I hit this error every other test run
```
>       np.testing.assert_array_equal(
            sc.metrics.morans_i(pbmc, obsm="X_pca"),
            sc.metrics.morans_i(g, pbmc.obsm["X_pca"].T),
        )
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 13 / 50 (26%)
E       Max absolute difference: 2.68156172e+154
E       Max relative difference: 1.
E        x: array([ 0.000000e+00, -1.363116e+57, -1.363116e+57, -1.363116e+57,
E              -1.363116e+57, -1.363116e+57, -1.363116e+57,  0.000000e+00,
E              -1.363116e+57, -1.363116e+57, -1.363116e+57, -1.363116e+57,...
E        y: array([ 0.000000e+000, -1.363116e+057, -1.363116e+057, -1.363116e+057,
E              -1.363116e+057, -1.363116e+057, -1.363116e+057,  0.000000e+000,
E              -1.363116e+057, -1.363116e+057, -1.363116e+057, -1.363116e+057,...
```

Reply:
This seems pretty strange, and I think is worth investigating more. Is it literally alternating test runs?

Reply:
yes, am testing now with types but I think only solution is to set float 32

Reply:
ok, only solution is to use np.float32 for the obsm_pca test. The other don't fail but I think it makes sense to keep it consistent. I am not really sure why that is, but yes it was failing 4/5 times the test run. tbh I see the point for using int64, but do we really need float 64? probably not i'd say no?

Reply:
So, the results you're showing in the above test are really wrong, so I think there's a larger issue at play here. I'll take a look into this.

But, I would say impression in these sorts of calculations can cause some pretty bad results. For example, once we get over a couple hundred thousand cells, variance calculation on float32 values can start accumulating significant amounts of imprecision, which can lead to very wrong differential expression results.

Reply:
So the values that you were getting out reminded me a lot of what was happening when I came across some numba parallelization issues: https://github.com/numba/numba/issues/6774#issuecomment-788789663 (the equivalent test was failing for `gearys_c`).

Looking into it, I think the same problem is occurring, it's just that your getting a consistent wrong answer with `float32`, but a inconsistent wrong answer with `float64`.

For example, on commit 33db5ced726996e292e6:

```python
import scanpy as sc
pbmc = sc.datasets.pbmc3k_processed()
sc.metrics.morans_i(pbmc.obsp["connectivities"], pbmc.obsm["X_pca"].T)
```

```
array([        0., -26711958., -26711958., -26711958.,         0.,
       -26711958., -26711958., -26711958.,         0., -26711958.,
       -26711958.,         0., -26711958., -26711958.,         0.,
       -26711958., -26711958.,         0., -26711958., -26711958.,
               0., -26711958., -26711958.,         0., -26711958.,
       -26711958.,         0., -26711958., -26711958.,         0.,
       -26711958., -26711958.,         0., -26711958., -26711958.,
               0., -26711958., -26711958.,         0., -26711958.,
       -26711958.,         0., -26711958., -26711958.,         0.,
       -26711958., -26711958.,         0., -26711958., -26711958.],
      dtype=float32)
```

This bug seems to be based on having nested parallelism and certain reductions. We can avoid it by just not having nested parallelism, which is what I've done for `gearys_c`.